crates/core/src/client/mod.rs

```diff
@@ -404,35 +404,23 @@ impl<P: Provider + Send + Sync> KakarotEthApi<P> for KakarotClient<P> {
         address: Address,
         contract_addresses: Vec<Address>,
     ) -> Result<TokenBalances, EthApiError<P::Error>> {
-        let entrypoint = FieldElement::from_byte_slice_be(&keccak256("balanceOf(address)").0[0..4])?;
-
         let addr: Felt252Wrapper = address.into();
         let addr: FieldElement = addr.into();
 
         let handles = contract_addresses.into_iter().map(|token_address| {
-            let calldata = vec![entrypoint, addr];
+            let token_addr: Felt252Wrapper = token_address.into();
+            let token = EthereumErc20::new(&self.kakarot_contract, token_addr.into());
 
-            self.call(
-                token_address,
-                Bytes::from(vec_felt_to_bytes(calldata).0),
-                BlockId::from(BlockNumberOrTag::Latest),
-            )
+            let block_id = StarknetBlockId::Tag(BlockTag::Latest);
+            FutureTokenBalance { balance: token.balance_of(addr, block_id), token_address }
         });
+
         let token_balances = join_all(handles)
             .await
             .into_iter()
-            .map(|token_address| match token_address {
-                Ok(call) => {
-                    let balance = U256::try_from_be_slice(call.as_ref())
-                        .ok_or(ConversionError::<()>::Other("error converting from Bytes to U256".into()))
-                        .unwrap();
-                    TokenBalance { contract_address: address, token_balance: Some(balance), error: None }
-                }
-                Err(e) => TokenBalance {
-                    contract_address: address,
-                    token_balance: None,
-                    error: Some(format!("kakarot_getTokenBalances Error: {e}")),
-                },
+            .map(|f| match f.0 {
```

Comment:
why name the closure argument f?

Reply:
Just called it f for future but can update :)

Reply:
also, can we use `f.balance` and `f.token_address` instead of `f.0` and `f.1`?

Reply:
I tried returning a structure but there are a couple of issues: returning a structure means I need to update the output produced by the `Future` implementation for `FutureTokenBalance`. I could for example set it to a `TokenBalance`. This means that balance needs to resolve to a U256, which again, I can do by setting it with `F: Future<Output = U256>`. The issue now is that actually the call to `token.balance_of(addr, block_id)` actually resolves to `Result<U256, P::Error>`. That would mean that the `balance` field for `FutureTokenBalance` needs to be fixed to `Future<Output = Result<U256, P::Error>>`. I can do this but I wasn't a huge fan of linking `FutureTokenBalance` so much to the provider. Let me know what you think.